File: model_1/CS3233_2025_Midterm/hungry_piplups_1_solution.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MAX_U = 1001000;

int next_ptr[MAX_U + 2];
int timestamp[MAX_U + 2];
int now = 0;

int find(int x, int U) {
    if (x > U) {
        return x;
    }
    vector<int> stack;
    int cur = x;
    while (cur <= U) {
        if (timestamp[cur] != now) {
            break;
        }
        if (next_ptr[cur] == cur) {
            break;
        }
        stack.push_back(cur);
        cur = next_ptr[cur];
    }
    if (cur <= U && timestamp[cur] != now) {
        timestamp[cur] = now;
        next_ptr[cur] = cur;
    }
    for (int node : stack) {
        next_ptr[node] = cur;
    }
    return cur;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    int N, M;
    cin >> N >> M;
    vector<int> A(N);
    for (int i = 0; i < N; i++) {
        cin >> A[i];
    }
    for (int i = 0; i < M; i++) {
        int type;
        cin >> type;
        if (type == 0) {
            int P, X;
            cin >> P >> X;
            A[P-1] = X;
        } else {
            int L, R, X_i;
            cin >> L >> R >> X_i;
            int len = R - L + 1;
            int U_bound = len + 1000;
            now++;
            for (int j = L-1; j <= R-1; j++) {
                int x = A[j];
                int y = find(x, U_bound);
                if (y <= U_bound) {
                    next_ptr[y] = find(y+1, U_bound);
                }
            }
            int current = 0;
            for (int cnt = 0; cnt < X_i; cnt++) {
                current = find(current+1, U_bound);
            }
            cout << current << '\n';
        }
    }
    return 0;
}

File: model_1/CS3233_2025_Midterm/hungry_piplups_1_solution.py
import sys

MAX_U = 1001000
next_ptr = list(range(MAX_U + 2))
timestamp = [0] * (MAX_U + 2)
now = 0

class FenwTree:
    def __init__(self, size):
        self.n = size
        self.tree = [0] * (self.n + 1)
    
    def update(self, index, delta):
        i = index + 1
        while i <= self.n:
            self.tree[i] += delta
            i += i & -i
            
    def query(self, index):
        if index < 0:
            return 0
        i = index + 1
        s = 0
        while i > 0:
            s += self.tree[i]
            i -= i & -i
        return s
        
    def range_sum(self, l, r):
        if l > r:
            return 0
        return self.query(r) - self.query(l - 1)

def find(x, U_bound):
    if x > U_bound:
        return x
    path = []
    cur = x
    while cur <= U_bound:
        if timestamp[cur] != now:
            break
        if next_ptr[cur] == cur:
            break
        path.append(cur)
        cur = next_ptr[cur]
    if cur > U_bound:
        for node in path:
            next_ptr[node] = cur
        return cur
    if timestamp[cur] != now:
        timestamp[cur] = now
        next_ptr[cur] = cur
    for node in path:
        next_ptr[node] = cur
    return cur

def main():
    global now
    data = sys.stdin.read().split()
    it = iter(data)
    n = int(next(it)); m = int(next(it))
    A = [0] * n
    for i in range(n):
        A[i] = int(next(it))
    
    fenw_trees = [None]
    for x in range(1, 501):
        fenw_trees.append(FenwTree(n))
    
    for i in range(n):
        if 1 <= A[i] <= 500:
            fenw_trees[A[i]].update(i, 1)
    
    out_lines = []
    for _ in range(m):
        t = int(next(it))
        if t == 0:
            p = int(next(it)); x_val = int(next(it))
            idx = p - 1
            old_val = A[idx]
            if 1 <= old_val <= 500:
                fenw_trees[old_val].update(idx, -1)
            if 1 <= x_val <= 500:
                fenw_trees[x_val].update(idx, 1)
            A[idx] = x_val
        else:
            L = int(next(it)); R = int(next(it)); X = int(next(it))
            len_seg = R - L + 1
            U_bound = len_seg + 1000
            now += 1
            freq = [0] * 501
            for x in range(1, 501):
                cnt = fenw_trees[x].range_sum(L - 1, R - 1)
                freq[x] = cnt
            for x in range(1, 501):
                cnt = freq[x]
                for _ in range(cnt):
                    y = find(x, U_bound)
                    if y <= U_bound:
                        next_ptr[y] = find(y + 1, U_bound)
            current = 0
            for j in range(X):
                current = find(current + 1, U_bound)
            out_lines.append(str(current))
    sys.stdout.write("\n".join(out_lines))

if __name__ == "__main__":
    main()